From 4222fd89769e8b03d4927c75695d3129daab1250 Mon Sep 17 00:00:00 2001
From: Hyun Kwon <hyun.kwon@xilinx.com>
Date: Mon, 25 Jun 2018 14:48:27 -0700
Subject: [PATCH 0382/1851] drm: xlnx: zynqmp_disp: Modul param for initial
 format of gfx layer

commit 2ea6f218fad8b406f57d576efe7ccb3438e116b3 from
https://github.com/Xilinx/linux-xlnx.git

This allows the initial format for the graphics layer to be specified
and changed through module param. The format is used for
the fbdev emulation, and the emulated fbdev will be initialized with
the specified format: 0 = rgb565, 1 = rgb888, 2 = argb8888.

Signed-off-by: Hyun Kwon <hyun.kwon@xilinx.com>
Reviewed-by: Satish Kumar Nagireddy <satish.nagireddy.nagireddy@xilinx.com>
Signed-off-by: Michal Simek <michal.simek@xilinx.com>
State: pending
Signed-off-by: Yaliang Wang <Yaliang.Wang@windriver.com>
---
 drivers/gpu/drm/xlnx/zynqmp_disp.c | 23 +++++++++++++++++++++--
 1 file changed, 21 insertions(+), 2 deletions(-)

diff --git a/drivers/gpu/drm/xlnx/zynqmp_disp.c b/drivers/gpu/drm/xlnx/zynqmp_disp.c
index 47a40bf83c19..0ed58a8881cd 100644
--- a/drivers/gpu/drm/xlnx/zynqmp_disp.c
+++ b/drivers/gpu/drm/xlnx/zynqmp_disp.c
@@ -70,6 +70,23 @@
  * - xlnx_crtc: Xilinx DRM specific crtc functions
  */
 
+/* The default value is ZYNQMP_DISP_AV_BUF_GFX_FMT_RGB565 */
+static uint zynqmp_disp_gfx_init_fmt;
+module_param_named(gfx_init_fmt, zynqmp_disp_gfx_init_fmt, uint, 0444);
+MODULE_PARM_DESC(gfx_init_fmt, "The initial format of the graphics layer\n"
+			       "\t\t0 = rgb565 (default)\n"
+			       "\t\t1 = rgb888\n"
+			       "\t\t2 = argb8888\n");
+/* These value should be mapped to index of av_buf_gfx_fmts[] */
+#define ZYNQMP_DISP_AV_BUF_GFX_FMT_RGB565		10
+#define ZYNQMP_DISP_AV_BUF_GFX_FMT_RGB888		5
+#define ZYNQMP_DISP_AV_BUF_GFX_FMT_ARGB8888		1
+static const u32 zynqmp_disp_gfx_init_fmts[] = {
+	ZYNQMP_DISP_AV_BUF_GFX_FMT_RGB565,
+	ZYNQMP_DISP_AV_BUF_GFX_FMT_RGB888,
+	ZYNQMP_DISP_AV_BUF_GFX_FMT_ARGB8888,
+};
+
 /* Blender registers */
 #define ZYNQMP_DISP_V_BLEND_BG_CLR_0			0x0
 #define ZYNQMP_DISP_V_BLEND_BG_CLR_1			0x4
@@ -1008,7 +1025,6 @@ static const struct zynqmp_disp_fmt av_buf_vid_fmts[] = {
 };
 
 /* List of graphics layer formats */
-#define ZYNQMP_DISP_AV_BUF_GFX_FMT_RGB565	10
 static const struct zynqmp_disp_fmt av_buf_gfx_fmts[] = {
 	{
 		.drm_fmt	= DRM_FORMAT_ABGR8888,
@@ -3083,6 +3099,7 @@ static int zynqmp_disp_enumerate_fmts(struct zynqmp_disp *disp)
 	struct zynqmp_disp_layer *layer;
 	u32 *bus_fmts;
 	u32 i, size, num_bus_fmts;
+	u32 gfx_fmt = ZYNQMP_DISP_AV_BUF_GFX_FMT_RGB565;
 
 	num_bus_fmts = ARRAY_SIZE(av_buf_live_fmts);
 	bus_fmts = devm_kzalloc(disp->dev, sizeof(*bus_fmts) * num_bus_fmts,
@@ -3119,7 +3136,9 @@ static int zynqmp_disp_enumerate_fmts(struct zynqmp_disp *disp)
 
 	for (i = 0; i < layer->num_fmts; i++)
 		layer->drm_fmts[i] = av_buf_gfx_fmts[i].drm_fmt;
-	layer->fmt = &av_buf_gfx_fmts[ZYNQMP_DISP_AV_BUF_GFX_FMT_RGB565];
+	if (zynqmp_disp_gfx_init_fmt < ARRAY_SIZE(zynqmp_disp_gfx_init_fmts))
+		gfx_fmt = zynqmp_disp_gfx_init_fmts[zynqmp_disp_gfx_init_fmt];
+	layer->fmt = &av_buf_gfx_fmts[gfx_fmt];
 
 	return 0;
 }
-- 
2.31.1

